### PR TITLE
ENH: Faster bandpass

### PR DESCRIPTION
**Pull request by danielballan**
@caspervdw I put the wining function from #219 into this PR. The simple test, which is based on what I did in the notebook, is failing. Feel free to take a look if you are inspired.

### CODE REVIEW

**Comment by caspervdw** [comment]
I will have a look. As the `gaussian_filter` is based on `ndimage.correlate1d` together with a Gaussian kernel generator, it will probably be even faster with that function. The Gaussian kernel can be generated using a memoizer.

**Comment by caspervdw** [comment]
I set up a PR to your branch @danielballan . The new tests are still failing because margin values differ. Because these are irrelevant to out tracking I propose slicing them off in the legacy tests.

Speed tests of the current pyfftw, your gaussian_filter implementation (`modified`) and my correlate1d implementation (`combined`):

    In [16]:
    %timeit -n 10 current_bandpass(next(small), 3, 11)
    %timeit -n 10 modified_bandpass(next(small), 3, 11)
    %timeit -n 10 combined_bandpass(next(small), 3, 11)
    10 loops, best of 3: 23.3 ms per loop
    10 loops, best of 3: 16.1 ms per loop
    10 loops, best of 3: 15.3 ms per loop
    In [17]:
    
    %timeit -n 10 current_bandpass(next(prime), 3, 11)
    %timeit -n 10 modified_bandpass(next(prime), 3, 11)
    %timeit -n 10 combined_bandpass(next(prime), 3, 11)
    10 loops, best of 3: 45.4 ms per loop
    10 loops, best of 3: 14.9 ms per loop
    10 loops, best of 3: 13.9 ms per loop
    In [18]:
    
    %timeit -n 10 current_bandpass(next(medium), 3, 11)
    %timeit -n 10 modified_bandpass(next(medium), 3, 11)
    %timeit -n 10 combined_bandpass(next(medium), 3, 11)
    10 loops, best of 3: 595 ms per loop
    10 loops, best of 3: 377 ms per loop
    10 loops, best of 3: 348 ms per loop

**Comment by nkeim** [comment]
Mentioning #142 here, which this will presumably close.

**Comment by caspervdw** [comment]
@danielballan I will be able to assist with this PR until Thursday, from then I will be on holiday until August 7th. (see the relevant PR on your branch https://github.com/danielballan/trackpy/pull/2)

**Comment by nkeim** [comment]
OK, this is fascinating: When you attempt to merge this into master (hopefully after merging fixes by @caspervdw and myself), you'll find that [`test_uncertainty_failure`](https://github.com/soft-matter/trackpy/pull/252) fails, and you won't know why. Turns out that the bandpass done in `uncertainty.roi()` changes just enough that it no longer leads to `ep == NaN` (instead, we get `ep == -0.0`!). I couldn't find a new set of parameters in `test_uncertainty_failure` to fix this. 

What does fix this test is to not have `uncertainty.roi()` do its own bandpass. (The test calls `locate()` with `preprocess=False`, because the test image is already built from gaussians.) 

So, 2 questions for discussion:

1. Why does this fix the problem? I.e. why the new `bandpass` behavior so different in this instance?

2. Why is `uncertainty.roi()` even doing a bandpass when `preprocess=False`? If the user calls `locate()` with `preprocess=False`, shouldn't trackpy presume that the image is already bandpassed? In the [discussion on uncertainty estimation](https://github.com/soft-matter/trackpy/pull/239), there was some mention of ways to accommodate users who do their own bandpass, but I can't see how the current code does. The present behavior instead acts as a performance "gotcha." Shouldn't `locate()` accept *both* `image` and `raw_image` when `preprocess=False`?

**Comment by danielballan** [comment]
@nkeim I merged the fixes, and it looks like Travis does not reproduce the failure of `test_uncertainty_failure`. The mystery deepens!

As I look at `roi`, I can't believe I ever thought that function signature made sense. It requires `image` but only uses it if nothing is passed to the optional argument `image_bandpassed`, in which case it does its own bandpass using hard-coded parameters.

Instead, there should only be one image argument, and it should be bandpassed before it goes in, either explicitly by the user or automatically by `locate()`. Sound better?

**Comment by nkeim** [comment]
A passing test?! Oh no!! :wink: 

Getting `-0.0` instead of `NaN` *would* be consistent with a mystery bug. But `test_uncertainty_failure` is not in this branch, only in master. Do you need to rebase this branch in order for Travis to run that test?

I agree with your proposed change to `roi`. Would `locate` then look like `locate(image, preprocess=True, raw_image=None, characterize=True)`? And maybe issue a warning if the user calls it with `preprocess=True, raw_image=None, characterize=True`, saying that `ep`, `signal`, and `raw_mass` may be incorrect?

There was the other idea tossed around in #239 to break down `locate()` into e.g. `preprocess()`, `find_peaks()`, `characterize()`, etc., so that the user could call `locate()` in most cases, but that it would be straightforward to build a functionally equivalent pipeline when needed. That is a little radical, but it does have its charms.

**Comment by nkeim** [comment]
If I may offer some cheerleading for trackpy v0.3: I dusted off our benchmarks to answer a question (that turned out to be moot). But look how far we've come from v0.2.4 to the tip of this branch!

```
asv compare --machine shortcake 5ec747d4470fe9decceb5d2de8b588983189f828 6f746fc02147f3f9bfc11cde6469ef862a580d18

All benchmarks:

    before     after       ratio
  [5       ] [6       ]
-  561.91ms   129.75ms      0.23  artificial.TimeSuite.time_locate_dense
-  499.00ms    26.67ms      0.05  artificial.TimeSuite.time_locate_sparse
      2.38s      2.72s      1.14  real_data.DenseFeatureSuite.time_numba
-     9.94s      4.75s      0.48  real_data.DenseLinkingSuite.time_numba
-     4.95s      2.20s      0.44  real_data.WaterFeatureSuite.time_numba
     18.91s     16.17s      0.85  real_data.WaterFeatureSuite.time_python
   347.52ms   190.18ms      0.55  real_data.WaterLinkingSuite.time_numba
   344.81ms   180.60ms      0.52  real_data.WaterLinkingSuite.time_python
```

The regression for `real_data.DenseFeatureSuite.time_numba` could use some attention — I'm going to give it a once-over before I drop this — but when you look at the gains…

(Note: I used numba 0.14 to put everyone on the same footing.)

**Comment by nkeim** [comment]
Update: #254 fixes the regression in `real_data.DenseFeatureSuite.time_numba`. I predict that the runtime will drop from 2.72 s to about 1.7 s. So, good news across the board, and the work remaining on this PR will be well worth it.

**Comment by danielballan** [comment]
@nkeim Are all of your compute servers dessert-themed?

This is very motivating! Before we settle this, I want to benchmark Casper's alternative myself. When I tried that approach  it didn't do quite as well as the one I proposed in this PR. Perhaps a slightly different implementation or a difference of architectures. I'll circle back to it once pims is tagged. Soon!

**Comment by tacaswell** [comment]
The dessert/food themed computer names is a trend from the Nagel lab (that I never picked up).

**Comment by caspervdw** [comment]
Good work on this! The benchmarks look promising.

I agree with your ideas about `uncertainty.roi()`, it should simply accept any image and assume that it is already bandpassed. I don't remember why I put the weird calling signature, good that you caught this before tagging 0.3.

A more modular approach of `locate()` would be great. This will hopefully produce more readable code and give a more flexible use. This however is a big refactoring and it should be done right. I'll make a separate issue to discuss this.

File: trackpy/preprocessing.py [hunk 1]
```diff
@@ -37,10 +100,16 @@ def ifftn(a):
...
+    """Remove noise and background variation.
+
+    Convolve with a Gaussian to remove short-wavelength noise and subtract out
+    long-wavelength variations, retaining features of intermediate scale.
+
+    This implementation performs a Fourier transform using numpy.
+    In benchmarks using typical inputs, it was found to be slower than the
+    ``bandpass`` function in this module.
+
     Parmeters
```
Comment (caspervdw) [inline trackpy/preprocessing.py]: typo

**Comment by danielballan** [comment]
Note: superseded by #259 and thus marked as merged by GH

**Pull request merged**